src/lib.rs

```diff
@@ -0,0 +1,166 @@
+
+use self::{
+    fetcher::fetch_prices,
+    interface::{PriceData, Serializable},
+};
+
+#[derive(Clone)]
+pub struct Configuration {
+    pub fetch_interval_seconds: u64,
+    pub contracts: Vec<(String, String)>,
+    pub provider: RootProvider<Http<Client>>,
+}
+
+/// ## Cryptoprices
+///
+/// Cryptoprices is a lightweight Rust library that provides your Rust applications with a direct
+/// link to the latest cryptocurrency prices. All data is retrieved from Chainlink decentralized
+/// price feeds. Just copy the contract addresses for the symbol that you would like to track from:
+/// https://data.chain.link/feeds.
+///
+///
+/// Note: Rustlink is designed to be ran on the main thread, in an asynchronous tokio environment
+#[derive(Clone)]
+pub struct CryptoPrices {
+    configuration: Configuration,
+    tree: Tree,
+}
+
+#[derive(Error, Debug)]
+pub enum DatabaseError {
+    #[error("Could not find symbol")]
+    NotFound,
+    #[error("Could not deserialize binary data")]
+    Deserialize,
+    #[error("Database internal error: {0}")]
+    SledError(#[from] sled::Error),
+}
+
+impl CryptoPrices {
+    /// Creates a new Rustlink instance.
+    ///
+    /// Expected parameters:
+    /// - `rpc_url`: The RPC url of your chosen EVM network where Chainlink offers decentralised data feeds.
+    /// Don't know where? Check https://data.chain.link/feeds.
+    /// - `fetch_interval_seconds`: How often to update prices in the database (to prevent RPC rate limitation)
+    /// - `sled_path`: The path for the database
+    /// - `contracts`: A tuple list containing a ticker name and its corresponding contract address on the
+    /// EVM chain.
+    ///
+    /// Example:
+    ///
+    /// ```rust
+    /// use cryptoprices::CryptoPrices;
+    /// let mut contracts:Vec<(String,String)>=Vec::new();
+    /// contracts.push(("BTC".to_string(),"0x9ef1B8c0E4F7dc8bF5719Ea496883DC6401d5b2e".to_string()));
+    /// let crypto_prices=CryptoPrices::new("https://bsc-dataseed1.binance.org/",10,"./cryptoprices",contracts);
+    /// ```
+    pub fn new(
+        rpc_url: &str,
+        fetch_interval_seconds: u64,
+        sled_path: &str,
+        contracts: Vec<(String, String)>,
+    ) -> CryptoPrices {
+        let db = sled::open(sled_path).unwrap();
+        let provider = ProviderBuilder::new().on_http(Url::from_str(rpc_url).unwrap());
+
+        // Setup logging
+        let logfile = FileAppender::builder()
+            .encoder(Box::new(PatternEncoder::new("{l} - {m}\n")))
+            .build("./rustlink.log")
+            .unwrap();
+
+        let config = Config::builder()
+            .appender(Appender::builder().build("logfile", Box::new(logfile)))
+            .build(Root::builder().appender("logfile").build(LevelFilter::Info))
+            .unwrap();
+
+        // Try to initialize and catch error silently if already initialized
+        // during tests this make this function throw error
+        if log4rs::init_config(config).is_err() {
```

Comment:
Not sure if it is a good idea to initialize a logger from within a library...  like tests should perhaps create their own logger?   Like I would hate if this somehow interferes with my env when I load this lib.

Reply:
Actually, this like totally freaks me out.  + `.gitignore` won't help here as logs will be created inside the folder of an application using this library.  I suggest either removing this completely or making logging optional via some configuration flag.  A local database or configuration files is understandable, but a library/crate creating its own logs without any external control over it is a rather bad etiquette. 

I would perhaps suggest accepting `Option<&Path>` that contains a log file name or if None, logging is skipped across the entire library.

Reply:
Good point, I had a custom logging mechanism before but was told that this is bad practice and that RUST_LOG env should be used.